File: src/main/java/org/jcodec/containers/mp4/boxes/FielExtension.java
package org.jcodec.containers.mp4.boxes;

import js.nio.ByteBuffer;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * @author The JCodec project
 * 
 */
public class FielExtension extends Box {
    public FielExtension(Header header) {
        super(header);
    }

    private int type;
    private int order;

    public static String fourcc() {
        return "fiel";
    }

    public boolean isInterlaced() {
        return type == 2;
    }

    public boolean topFieldFirst() {
        return order == 1 || order == 6;
    }

    public String getOrderInterpretation() {
        if (isInterlaced())
            // Copy from qtff 2007-09-04, page 98 The following defines
            // the permitted variants:
            // 0 There is only one field.
            switch (order) {
            case 1:
                // 1 T is displayed earliest, T is stored first in the file.
                return "top";
            case 6:
                // 6 B is displayed earliest, B is stored first in the file.
                return "bottom";
            case 9:
                // 9  B is displayed earliest, T is stored first in the file.
                return "bottomtop";
            case 14:
                // 14  T is displayed earliest, B is stored first in the
                // file.
                return "topbottom";
            }

        return "";
    }

    @Override
    public void parse(ByteBuffer input) {
        this.type = input.get() & 0xff;
        if (isInterlaced()) {
            this.order = input.get() & 0xff;
        }
    }

    @Override
    public void doWrite(ByteBuffer out) {
        out.put((byte) type);
        out.put((byte) order);
    }
    
    @Override
    public int estimateSize() {
        return 2 + 8;
    }
}

File: src/main/java/org/jcodec/containers/mp4/boxes/FielExtension.kt
package org.jcodec.containers.mp4.boxes

import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * @author The JCodec project
 */
class FielExtension(header: Header) : Box(header) {
    private var type = 0
    private var order = 0
    val isInterlaced: Boolean
        get() = type == 2

    fun topFieldFirst(): Boolean {
        return order == 1 || order == 6
    }// 14  T is displayed earliest, B is stored first in the
    // file.
// 9  B is displayed earliest, T is stored first in the file.// 6 B is displayed earliest, B is stored first in the file.

    // 1 T is displayed earliest, T is stored first in the file.
    val orderInterpretation: String
        get() {
            if (isInterlaced) when (order) {
                1 ->                 // 1 T is displayed earliest, T is stored first in the file.
                    return "top"
                6 ->                 // 6 B is displayed earliest, B is stored first in the file.
                    return "bottom"
                9 ->                 // 9  B is displayed earliest, T is stored first in the file.
                    return "bottomtop"
                14 ->                 // 14  T is displayed earliest, B is stored first in the
                    // file.
                    return "topbottom"
            }
            return ""
        }

    override fun parse(input: ByteBuffer) {
        type = input.get().toInt() and 0xff
        if (isInterlaced) {
            order = input.get().toInt() and 0xff
        }
    }

    public override fun doWrite(out: ByteBuffer) {
        out.put(type.toByte())
        out.put(order.toByte())
    }

    override fun estimateSize(): Int {
        return 2 + 8
    }

    companion object {
        @JvmStatic
        fun fourcc(): String {
            return "fiel"
        }
    }
}